# **Tarea Nº 3**
Scope: If, Loops, Functions

Curso:          Fundamentos de Programación en Python para Macroeconomía y Finanzas     
Profesor:       Victor Fuentes Campos     

- Fecha y hora de entrega: Miércoles 10 de enero, 11:59am
- Agregue los nombres de las personas con las que discutió esta tarea: Adriana Arce y Braulio Mundaca
- Envíe su tarea haciendo el `push` de su código a su repo en GitHub Classroom: 

As of 2022 Q2, there are almost 3.5 million Andriod apps on Google Play and 1.6 million iOS apps on the App Store, and counting. Giving the digital revolution, there is an increase interest in investment opportunities in this sector

<img src="popular_app_stores.png" max-width="300">

Under this context, you will assess a financial committee interested in investing in software development. The goal is to determine the kinds of apps that are likely to attract more users because the number of people using our apps affect our revenue. For this purpose, you will access to data from Google Play and App Store.

- Google Play: you can dowload the dataset containing approximately 11k Android apps from [this link](https://dq-content.s3.amazonaws.com/350/googleplaystore.csv). It was collected in August 2018
- App Store: you can dowload the dataset containing roughly 7k iOS apps from [this link](https://dq-content.s3.amazonaws.com/350/AppleStore.csv). It was collected in July 2017

## Exploring data

1. Open the two datasets we mentioned above, and save both as lists of lists

In [2]:
# Solution
import pandas as pd

# Abrimos el primer conjunto de datos y lo guarda como una lista de listas
Google_Play = pd.read_csv(r"googleplaystore.csv") #(Colocar dirección del archivo csv)
dataset1 = Google_Play.values.tolist()

# Abrimos el segundo conjunto de datos y lo guarda como una lista de listas
Apple_Store = pd.read_csv(r"AppleStore.csv") #(Colocar dirección del archivo csv)
dataset2 = Apple_Store.values.tolist()

#print(dataset1, dataset2), quitarle el numeral en caso se quiera ver las listas

2. Print the first few rows of each dataset and find the number of rows and columns of each dataset

In [3]:
# Solution (1ra parte)

# Imprimimos las primeras filas del primer conjunto de datos
Google_Play.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  Reviews  Size     Installs  Type Price Content Rating  \
0     159   19M      10,000+  Free     0       Everyone   
1     967   14M     500,000+  Free     0       Everyone   
2   87510  8.7M   5,000,000+  Free     0       Everyone   
3  215644   25M  50,000,000+  Free     0           Teen   
4     967  2.8M     100,000+  Free     0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

In [4]:
# Encontramos e imprimimos el número de filas y columnas del primer conjunto de datos
print(f'El primer conjunto de datos tiene {Google_Play.shape[0]} filas y {Google_Play.shape[1]} columnas.') #Aquí el [0] hace referencia a las filas y [1] hace referencia a las columnas

El primer conjunto de datos tiene 10841 filas y 13 columnas.


In [5]:
#Solution(2da parte)

# Imprime las primeras filas del segundo conjunto de datos
Apple_Store.head()

id               track_name  size_bytes currency  price  \
0  284882215                 Facebook   389879808      USD    0.0   
1  389801252                Instagram   113954816      USD    0.0   
2  529479190           Clash of Clans   116476928      USD    0.0   
3  420009108               Temple Run    65921024      USD    0.0   
4  284035177  Pandora - Music & Radio   130242560      USD    0.0   

   rating_count_tot  rating_count_ver  user_rating  user_rating_ver      ver  \
0           2974676               212          3.5              3.5     95.0   
1           2161558              1289          4.5              4.0    10.23   
2           2130805               579          4.5              4.5  9.24.12   
3           1724546              3842          4.5              4.0    1.6.2   
4           1126879              3594          4.0              4.5    8.4.1   

  cont_rating        prime_genre  sup_devices.num  ipadSc_urls.num  lang.num  \
0          4+  Social Networking               37                1        29   
1         12+      Photo & Video               37                0        29   
2          9+              Games               38                5        18   
3          9+              Games               40                5         1   
4         12+              Music               37                4         1   

   vpp_lic  
0        1  
1        1  
2        1  
3        1  
4        1

In [6]:
# Encuentra e imprime el número de filas y columnas del segundo conjunto de datos
print(f'El segundo conjunto de datos tiene {Apple_Store.shape[0]} filas y {Apple_Store.shape[1]} columnas.')

El segundo conjunto de datos tiene 7197 filas y 16 columnas.


For more info about the columns in the App Store dataset, visit [this link](https://www.kaggle.com/datasets/ramamet4/app-store-apple-data-set-10k-apps?select=AppleStore.csv)

## Preparing data

3. As noted before in the [discussion forum](https://www.kaggle.com/datasets/lava18/google-play-store-apps/discussion/66015), there is an error in the Google Play dataset. Please, remove the entry 10472 and explain shortly why

In [7]:
# Solution
print(Google_Play.iloc[10472])
print(dataset1[10472])

App               Life Made WI-Fi Touchscreen Photo Frame
Category                                              1.9
Rating                                               19.0
Reviews                                              3.0M
Size                                               1,000+
Installs                                             Free
Type                                                    0
Price                                            Everyone
Content Rating                                        NaN
Genres                                  February 11, 2018
Last Updated                                       1.0.19
Current Ver                                    4.0 and up
Android Ver                                           NaN
Name: 10472, dtype: object
['Life Made WI-Fi Touchscreen Photo Frame', '1.9', 19.0, '3.0M', '1,000+', 'Free', '0', 'Everyone', nan, 'February 11, 2018', '1.0.19', '4.0 and up', nan]


In [8]:
#Vamos a contar el número de columna de la fila 10472
fila_10472= Google_Play.loc[10472]
num_columnas = len(fila_10472)
print(f'La fila 10472 tiene {num_columnas} columnas.')

La fila 10472 tiene 13 columnas.


In [9]:
Google_Play_sin_10472=Google_Play.drop(10472)
Google_Play_sin_10472.iloc[10472]
#Ahora si podemos verificar que la fila 10472 se ha eliminado ya que cuando buscamos nuevamente la fila tenemos diferentes valores.

App               osmino Wi-Fi: free WiFi
Category                            TOOLS
Rating                                4.2
Reviews                            134203
Size                                 4.1M
Installs                      10,000,000+
Type                                 Free
Price                                   0
Content Rating                   Everyone
Genres                              Tools
Last Updated               August 7, 2018
Current Ver                       6.06.14
Android Ver                    4.4 and up
Name: 10473, dtype: object

In [10]:
#Respuesta: A diferencia de otras filas, en la fila 10472 hay 2 entradas con missing values, lo cual podría responder a un mal registro de la app.

4. Count the number of duplicates in the Google Play dataset. You must build a function to loop through the Google Play dataset

In [11]:
# Creamos la función
def contar_duplicados(Google_Play_sin_10472):
    """ Esta función cuenta los duplicados de la base de datos Google_Play_sin_10472 """
    total_filas=len(Google_Play_sin_10472)
    filas_únicas=len(Google_Play_sin_10472.drop_duplicates())
    filas_duplicadas = total_filas - filas_únicas
    return filas_duplicadas 

#Usamos la función
duplicados = contar_duplicados(Google_Play_sin_10472)
print(f'Hay {duplicados} filas duplicadas en el dataset de Google Play.')

Hay 483 filas duplicadas en el dataset de Google Play.


5. Remove duplicates ...

In [12]:
#Creamos una función
def eliminar_duplicados(Google_Play_sin_10472):
    """ Esta función elimina los duplicados de la base de datos Google_Play_sin_10472 """
    Google_Play_sin_duplicados = Google_Play_sin_10472.drop_duplicates()
    return Google_Play_sin_duplicados

#Usamos la función
Google_Play_sin_duplicados=eliminar_duplicados(Google_Play_sin_10472)

In [13]:
#Para verificar que se han eliminado, volvemos a usar la función de contar duplicados
duplicados = contar_duplicados(Google_Play_sin_duplicados)
print(f'Hay {duplicados} filas duplicadas en el dataset de Google Play.')

Hay 0 filas duplicadas en el dataset de Google Play.


5. a. Create a dictionary, where each dictionary key is a unique app name and the corresponding dictionary value is the highest number of reviews of that app


In [14]:
# Solution
#Definimos la funcion
def create_app_reviews_dictionary(dataframe):
    """Usaremos la base de datos Google_Play_sin_duplicados"""
    app_reviews_dict = {}

    for index, row in dataframe.iterrows():
        app_name = row['App']
        app_reviews = row['Reviews']

        if app_name not in app_reviews_dict or app_reviews > app_reviews_dict[app_name]:
            app_reviews_dict[app_name] = app_reviews

    return app_reviews_dict
result_dict = create_app_reviews_dictionary(Google_Play_sin_duplicados)

5. b. Use the information stored in the dictionary and create a new dataset, which will have only one entry per app (and for each app, we'll only select the entry with the highest number of reviews).


In [15]:
# Solution
Google_Play_filtered_dataframe = (Google_Play_sin_duplicados.sort_values(by='Reviews', ascending=False)
                      .drop_duplicates('App')
                      .reset_index(drop=True))

5. c. Explore the cleaned dataset, it should have 9,659 rows

In [16]:
# Solution
print(Google_Play_filtered_dataframe)

                                                    App          Category  \
0                                GollerCepte Live Score            SPORTS   
1                          Ad Block REMOVER - NEED ROOT             TOOLS   
2                                   SnipSnap Coupon App          SHOPPING   
3                     US Open Tennis Championships 2018            SPORTS   
4                                            DreamTrips  TRAVEL_AND_LOCAL   
...                                                 ...               ...   
9654  Curso Básico y Avanzado la Fe de Jesús Audio-T...            FAMILY   
9655                           CT Pulmonary Angiography           MEDICAL   
9656                           Theme For Galaxy Note FE   PERSONALIZATION   
9657                                     Toronto Dating            DATING   
9658                               PlayMotiv dn edition          BUSINESS   

      Rating Reviews  Size    Installs  Type Price Content Rating  \
0     

6. Remove non-english apps from both Android and Apple datasets. You should write a function that takes a string, if there's more than three characters that fall outside the ASCII range (0 - 127), then the function should return False (identify the string as non-English), otherwise it should return True. Hint: use the [```ord()```  function](https://www.w3schools.com/python/ref_func_ord.asp)

In [17]:
# For instance
correct_english_name = "Instachat 😜"  # 1 character outside the ASCII range (😜)
incorrect_english_name = "爱奇艺PPS -《欢乐颂2》电视剧热播" # 13 character outside the ASCII range (ie 播)

In [18]:
print(ord("I"))
print(ord("t"))
print(ord("-"))
print(ord("😜"))
print(ord("爱"))

73
116
45
128540
29233


In [19]:
# Solution
import pandas as pd

#Primero, vamos a identificar el nombre de la columna que contiene los nombres de las aplicaciones en ambas bases de datos
#Caso de Google Play:
Google_Play_filtered_dataframe.head(0) #App


Empty DataFrame
Columns: [App, Category, Rating, Reviews, Size, Installs, Type, Price, Content Rating, Genres, Last Updated, Current Ver, Android Ver]
Index: []

In [20]:
#Caso de Apple Store:
Apple_Store.head(0) #track_name

Empty DataFrame
Columns: [id, track_name, size_bytes, currency, price, rating_count_tot, rating_count_ver, user_rating, user_rating_ver, ver, cont_rating, prime_genre, sup_devices.num, ipadSc_urls.num, lang.num, vpp_lic]
Index: []

In [21]:
#Ahora si pasaremos a definir la función

#Caso de Google Play:
def identificar_apps_en_ingles(oogle_Play_filtered_dataframe, App):
    """ Esta función identifica si los nombres de las aplicaciones en la base de datos de Google Play son en inglés o no. 
    Devuelve una nueva base de datos con una columna adicional que indica si el nombre de cada aplicación es en inglés o no."""
    def en_ingles(App):
        caracteres_no_ascii = 0
        for caracter in App:
            if ord(caracter) > 127:
                caracteres_no_ascii += 1
        if caracteres_no_ascii > 3:
            return False
        else:
            return True

    Google_Play_filtered_dataframe.loc[:, 'En_ingles'] = Google_Play_filtered_dataframe[App].apply(en_ingles)
    return Google_Play_filtered_dataframe
#Usamos la función:
nombres_de_aplicaciones_columna = 'App' 
Google_Play_apps = identificar_apps_en_ingles(Google_Play_filtered_dataframe, nombres_de_aplicaciones_columna)


Google_Play_apps


App          Category  \
0                                GollerCepte Live Score            SPORTS   
1                          Ad Block REMOVER - NEED ROOT             TOOLS   
2                                   SnipSnap Coupon App          SHOPPING   
3                     US Open Tennis Championships 2018            SPORTS   
4                                            DreamTrips  TRAVEL_AND_LOCAL   
...                                                 ...               ...   
9654  Curso Básico y Avanzado la Fe de Jesús Audio-T...            FAMILY   
9655                           CT Pulmonary Angiography           MEDICAL   
9656                           Theme For Galaxy Note FE   PERSONALIZATION   
9657                                     Toronto Dating            DATING   
9658                               PlayMotiv dn edition          BUSINESS   

      Rating Reviews  Size    Installs  Type Price Content Rating  \
0        4.2    9992   31M  1,000,000+  Free     0       Everyone   
1        3.3     999   91k    100,000+  Free     0       Everyone   
2        4.2    9975   18M  1,000,000+  Free     0       Everyone   
3        4.0    9971   33M  1,000,000+  Free     0       Everyone   
4        4.7    9971   22M    500,000+  Free     0           Teen   
...      ...     ...   ...         ...   ...   ...            ...   
9654     NaN       0   11M         10+  Free     0   Everyone 10+   
9655     NaN       0  4.9M        100+  Free     0       Everyone   
9656     NaN       0  7.5M        100+  Free     0       Everyone   
9657     NaN       0   14M        100+  Free     0     Mature 17+   
9658     NaN       0   28M        500+  Free     0       Everyone   

               Genres       Last Updated Current Ver Android Ver  En_ingles  
0              Sports       May 23, 2018         6.5  4.1 and up       True  
1               Tools  December 17, 2013         3.2  2.2 and up       True  
2            Shopping   January 22, 2018         1.4  4.3 and up       True  
3              Sports       June 5, 2018         7.1  5.0 and up       True  
4      Travel & Local     August 6, 2018      1.28.1  5.0 and up       True  
...               ...                ...         ...         ...        ...  
9654        Education      July 12, 2018         1.0  4.0 and up       True  
9655          Medical     March 24, 2018         1.0  4.0 and up       True  
9656  Personalization     April 19, 2018         1.1  4.1 and up       True  
9657           Dating      July 12, 2018         1.0  4.1 and up       True  
9658         Business      April 8, 2018       1.0.2  4.1 and up       True  

[9659 rows x 14 columns]

In [22]:
#Caso de Apple Store:
def identificar_apps_en_ingles(Apple_Store, track_name):
    """ Esta función identifica si los nombres de las aplicaciones en la base de datos de Google Play son en inglés o no. 
    Devuelve una nueva base de datos con una columna adicional que indica si el nombre de cada aplicación es en inglés o no."""
    def en_ingles(track_name):
        caracteres_no_ascii = 0
        for caracter in track_name:
            if ord(caracter) > 127:
                caracteres_no_ascii += 1
        if caracteres_no_ascii > 3:
            return False
        else:
            return True

    Apple_Store.loc[:, 'En_ingles'] = Apple_Store[track_name].apply(en_ingles)
    return Apple_Store
#Usamos la función:
nombres_de_aplicaciones_columna = 'track_name' 
Apple_Store_apps = identificar_apps_en_ingles(Apple_Store, nombres_de_aplicaciones_columna)

Apple_Store_apps

id                                    track_name  size_bytes  \
0      284882215                                      Facebook   389879808   
1      389801252                                     Instagram   113954816   
2      529479190                                Clash of Clans   116476928   
3      420009108                                    Temple Run    65921024   
4      284035177                       Pandora - Music & Radio   130242560   
...          ...                                           ...         ...   
7192  1170406182  Shark Boom - Challenge Friends with your Pet   245415936   
7193  1069830936                           【謎解き】ヤミすぎ彼女からのメッセージ    16808960   
7194  1070052833                                    Go!Go!Cat!    91468800   
7195  1081295232   Suppin Detective: Expose their true visage!    83026944   
7196   977965019           みんなのお弁当 by クックパッド ~お弁当をレシピ付きで記録・共有~    51174400   

     currency  price  rating_count_tot  rating_count_ver  user_rating  \
0         USD    0.0           2974676               212          3.5   
1         USD    0.0           2161558              1289          4.5   
2         USD    0.0           2130805               579          4.5   
3         USD    0.0           1724546              3842          4.5   
4         USD    0.0           1126879              3594          4.0   
...       ...    ...               ...               ...          ...   
7192      USD    0.0                 0                 0          0.0   
7193      USD    0.0                 0                 0          0.0   
7194      USD    0.0                 0                 0          0.0   
7195      USD    0.0                 0                 0          0.0   
7196      USD    0.0                 0                 0          0.0   

      user_rating_ver      ver cont_rating        prime_genre  \
0                 3.5     95.0          4+  Social Networking   
1                 4.0    10.23         12+      Photo & Video   
2                 4.5  9.24.12          9+              Games   
3                 4.0    1.6.2          9+              Games   
4                 4.5    8.4.1         12+              Music   
...               ...      ...         ...                ...   
7192              0.0    1.0.9          4+              Games   
7193              0.0      1.2          9+               Book   
7194              0.0    1.1.2         12+              Games   
7195              0.0    1.0.3         12+      Entertainment   
7196              0.0    1.4.0          4+       Food & Drink   

      sup_devices.num  ipadSc_urls.num  lang.num  vpp_lic  En_ingles  
0                  37                1        29        1       True  
1                  37                0        29        1       True  
2                  38                5        18        1       True  
3                  40                5         1        1       True  
4                  37                4         1        1       True  
...               ...              ...       ...      ...        ...  
7192               38                5         1        1       True  
7193               38                0         1        1      False  
7194               37                2         2        1       True  
7195               40                0         1        1       True  
7196               37                0         1        1      False  

[7197 rows x 17 columns]

In [23]:
#Ahora, eliminaremos las aplicaciones que no estén en inglés de las bases de datos de Google Play y App Store.

#Caso Google Play: Creamos una nueva base de datos que solo contiene las aplicaciones en inglés
Google_Play_ingles = Google_Play_apps[Google_Play_apps['En_ingles'] == True]
#Caso App Store: Creamos una nueva base de datos que solo contiene las aplicaciones en inglés
Apple_Store_ingles = Apple_Store_apps[Apple_Store_apps['En_ingles'] == True]

In [24]:
#Visualizamos:
Google_Play_ingles

App          Category  \
0                                GollerCepte Live Score            SPORTS   
1                          Ad Block REMOVER - NEED ROOT             TOOLS   
2                                   SnipSnap Coupon App          SHOPPING   
3                     US Open Tennis Championships 2018            SPORTS   
4                                            DreamTrips  TRAVEL_AND_LOCAL   
...                                                 ...               ...   
9654  Curso Básico y Avanzado la Fe de Jesús Audio-T...            FAMILY   
9655                           CT Pulmonary Angiography           MEDICAL   
9656                           Theme For Galaxy Note FE   PERSONALIZATION   
9657                                     Toronto Dating            DATING   
9658                               PlayMotiv dn edition          BUSINESS   

      Rating Reviews  Size    Installs  Type Price Content Rating  \
0        4.2    9992   31M  1,000,000+  Free     0       Everyone   
1        3.3     999   91k    100,000+  Free     0       Everyone   
2        4.2    9975   18M  1,000,000+  Free     0       Everyone   
3        4.0    9971   33M  1,000,000+  Free     0       Everyone   
4        4.7    9971   22M    500,000+  Free     0           Teen   
...      ...     ...   ...         ...   ...   ...            ...   
9654     NaN       0   11M         10+  Free     0   Everyone 10+   
9655     NaN       0  4.9M        100+  Free     0       Everyone   
9656     NaN       0  7.5M        100+  Free     0       Everyone   
9657     NaN       0   14M        100+  Free     0     Mature 17+   
9658     NaN       0   28M        500+  Free     0       Everyone   

               Genres       Last Updated Current Ver Android Ver  En_ingles  
0              Sports       May 23, 2018         6.5  4.1 and up       True  
1               Tools  December 17, 2013         3.2  2.2 and up       True  
2            Shopping   January 22, 2018         1.4  4.3 and up       True  
3              Sports       June 5, 2018         7.1  5.0 and up       True  
4      Travel & Local     August 6, 2018      1.28.1  5.0 and up       True  
...               ...                ...         ...         ...        ...  
9654        Education      July 12, 2018         1.0  4.0 and up       True  
9655          Medical     March 24, 2018         1.0  4.0 and up       True  
9656  Personalization     April 19, 2018         1.1  4.1 and up       True  
9657           Dating      July 12, 2018         1.0  4.1 and up       True  
9658         Business      April 8, 2018       1.0.2  4.1 and up       True  

[9614 rows x 14 columns]

In [25]:
#Visualizamos
Apple_Store_ingles

id                                    track_name  size_bytes  \
0      284882215                                      Facebook   389879808   
1      389801252                                     Instagram   113954816   
2      529479190                                Clash of Clans   116476928   
3      420009108                                    Temple Run    65921024   
4      284035177                       Pandora - Music & Radio   130242560   
...          ...                                           ...         ...   
7182  1070854722                                  Be-be-bears!   480781312   
7186  1169971902                   Hey Duggee: We Love Animals   136347648   
7192  1170406182  Shark Boom - Challenge Friends with your Pet   245415936   
7194  1070052833                                    Go!Go!Cat!    91468800   
7195  1081295232   Suppin Detective: Expose their true visage!    83026944   

     currency  price  rating_count_tot  rating_count_ver  user_rating  \
0         USD   0.00           2974676               212          3.5   
1         USD   0.00           2161558              1289          4.5   
2         USD   0.00           2130805               579          4.5   
3         USD   0.00           1724546              3842          4.5   
4         USD   0.00           1126879              3594          4.0   
...       ...    ...               ...               ...          ...   
7182      USD   2.99                 0                 0          0.0   
7186      USD   2.99                 0                 0          0.0   
7192      USD   0.00                 0                 0          0.0   
7194      USD   0.00                 0                 0          0.0   
7195      USD   0.00                 0                 0          0.0   

      user_rating_ver      ver cont_rating        prime_genre  \
0                 3.5     95.0          4+  Social Networking   
1                 4.0    10.23         12+      Photo & Video   
2                 4.5  9.24.12          9+              Games   
3                 4.0    1.6.2          9+              Games   
4                 4.5    8.4.1         12+              Music   
...               ...      ...         ...                ...   
7182              0.0  1.0.2.5          4+              Games   
7186              0.0      1.2          4+              Games   
7192              0.0    1.0.9          4+              Games   
7194              0.0    1.1.2         12+              Games   
7195              0.0    1.0.3         12+      Entertainment   

      sup_devices.num  ipadSc_urls.num  lang.num  vpp_lic  En_ingles  
0                  37                1        29        1       True  
1                  37                0        29        1       True  
2                  38                5        18        1       True  
3                  40                5         1        1       True  
4                  37                4         1        1       True  
...               ...              ...       ...      ...        ...  
7182               35                5        13        1       True  
7186               40                5         1        1       True  
7192               38                5         1        1       True  
7194               37                2         2        1       True  
7195               40                0         1        1       True  

[6183 rows x 17 columns]

7. Loop through each dataset to isolate the free apps in separate lists. Make sure you identify the columns describing the app price correctly. Be careful, prices come up as strings ('0', $0.99, $2.99, etc.), so make sure you're not checking an integer or a float in your conditional statements. How many apps are free?


In [26]:
# Solution
import pandas as pd

#Primero debemos identificar la columna en la que se colocan los precios de las aplicaciones
#Caso de Google Play:
Google_Play_ingles.head(0) #Price

Empty DataFrame
Columns: [App, Category, Rating, Reviews, Size, Installs, Type, Price, Content Rating, Genres, Last Updated, Current Ver, Android Ver, En_ingles]
Index: []

In [27]:
#Caso de Apple Store:
Apple_Store_ingles.head(0) #price (este no tiene mayúscula)

Empty DataFrame
Columns: [id, track_name, size_bytes, currency, price, rating_count_tot, rating_count_ver, user_rating, user_rating_ver, ver, cont_rating, prime_genre, sup_devices.num, ipadSc_urls.num, lang.num, vpp_lic, En_ingles]
Index: []

In [37]:
# Caso Google Play:
# Creamos un nuevo DataFrame que solo contiene las aplicaciones gratuitas
apps_gratuitas_Google_Play = Google_Play_ingles.loc[Google_Play_ingles['Price'] == '0']
#Creamos una lista de las aplicaciones gratuitas
lista_apps_gratuitas_Google_Play=apps_gratuitas_Google_Play['App'].tolist()
print(lista_apps_gratuitas_Google_Play)

['GollerCepte Live Score', 'Ad Block REMOVER - NEED ROOT', 'SnipSnap Coupon App', 'US Open Tennis Championships 2018', 'DreamTrips', 'Adult Color by Number Book - Paint Mandala Pages', 'BSPlayer ARMv7 VFP CPU support', 'Easy Resume Builder, Resume help, Curriculum vitae', 'MegaFon Dashboard', 'Al jazeera TV', 'Buff Thun - Daily Free Webtoon / Comics / Web Fiction / Mini Game', 'Rope Hero Return of a Legend', 'Family Guy The Quest for Stuff', 'tutti.ch - Free Classifieds', 'Teach Me Anatomy', 'Funny Pics', 'Files To SD Card Pro', 'Best Car Wallpapers', 'HSPA+ Tweaker (3G booster)', 'Art Pixel Coloring. Color by Number.', 'Adblock Fast', 'Zombie Catchers', 'Udemy - Online Courses', 'Tennis 24 - tennis live scores', 'Z-Empire: Dead Strike', 'Kids Craft Ideas', 'AKASO DV', 'CS Interview FAQs', 'PTI Flex Maker, Photo Frame Editor & Songs 2018', 'Gang Wars of San Andreas', 'Display Phone Screen On TV', 'FreedomPop Messaging Phone/SIM', 'Wrestling Revolution 3D', 'English words application mi

In [29]:
# Imprimimos el número de aplicaciones gratuitas
print(f"En el caso de Google Play, hay {len(apps_gratuitas_Google_Play)} aplicaciones gratuitas.")

En el caso de Google Play, hay 8862 aplicaciones gratuitas.


In [39]:
# Caso Apple Store:
# Creamos un nuevo DataFrame que solo contiene las aplicaciones gratuitas
apps_gratuitas_Apple_Store = Apple_Store_ingles.loc[Apple_Store_ingles['price'] == 0]
#Creamos una lista de las aplicaciones gratuitas
lista_apps_gratuitas_Apple_Store=apps_gratuitas_Apple_Store['track_name'].tolist()
print(lista_apps_gratuitas_Apple_Store)

['Facebook', 'Instagram', 'Clash of Clans', 'Temple Run', 'Pandora - Music & Radio', 'Pinterest', 'Bible', 'Candy Crush Saga', 'Spotify Music', 'Angry Birds', 'Subway Surfers', 'Solitaire', 'CSR Racing', 'Crossy Road - Endless Arcade Hopper', 'Injustice: Gods Among Us', 'Hay Day', 'PAC-MAN', 'Calorie Counter & Diet Tracker by MyFitnessPal', 'DragonVale', 'The Weather Channel: Forecast, Radar & Alerts', 'Head Soccer', 'Google – Search made just for mobile', 'Despicable Me: Minion Rush', 'The Sims™ FreePlay', 'Google Earth', 'Sonic Dash', 'Groupon - Deals, Coupons & Discount Shopping App', '8 Ball Pool™', 'Tiny Tower - Free City Building', 'Jetpack Joyride', 'Bike Race - Top Motorcycle Racing Games', 'Shazam - Discover music, artists, videos & lyrics', 'Kim Kardashian: Hollywood', 'Trivia Crack', 'WordBrain', 'Sniper 3D Assassin: Shoot to Kill Gun Game', 'Flow Free', 'Lose It! – Weight Loss Program and Calorie Counter', 'Skype for iPhone', 'Geometry Dash Lite', '▻Sudoku', 'Twitter', 'Mes

In [40]:
# Imprimimos el número de aplicaciones gratuitas
print(f"En el caso de Apple Store, hay {len(apps_gratuitas_Apple_Store)} aplicaciones gratuitas.")

En el caso de Apple Store, hay 3222 aplicaciones gratuitas.


## Exploring data

 8. Build a frequency table for the ```prime_genre``` column of the App Store dataset, and the ```Genres``` and ```Category``` columns of the Google Play data set. What are the most common genres for each market? You must develop and deploy your function

In [31]:
# Solution
#Definimos nuestra funcion usando las bases de datos originales
def frequency_table(dataframe, column_name):
    return dataframe[column_name].value_counts().reset_index().rename(columns={'index': column_name, column_name: 'Frequency'})

# Frecuencias de 'prime_genre'
app_store_frequency = frequency_table(Apple_Store, 'prime_genre')

# Frecuencias de 'Genres'
google_play_genres_frequency = frequency_table(Google_Play, 'Genres')

# Frecuencias de 'Category'
google_play_category_frequency = frequency_table(Google_Play, 'Category')
print(app_store_frequency.head())
print(google_play_genres_frequency.head())
print(google_play_category_frequency.head())

       Frequency  count
0          Games   3862
1  Entertainment    535
2      Education    453
3  Photo & Video    349
4      Utilities    248
       Frequency  count
0          Tools    842
1  Entertainment    623
2      Education    549
3        Medical    463
4       Business    460
  Frequency  count
0    FAMILY   1972
1      GAME   1144
2     TOOLS    843
3   MEDICAL    463
4  BUSINESS    460


 9. One way to find out what genres are the most popular (have the most users) is to calculate the average number of installs for each app genre. Based on that exercise, give one app profile recommendation for Google Play.
 
    For the Google Play data set, we can find the required information in the ```Installs``` column. Then, compute the average number of installs by genre. Hint 1: consider only unique values of the ```Category``` column. Hint 2: If an app has, for instance, ```1,000,000+``` assume it as ```1,000,000``` installs. Hint 3: Be careful with a skewed distribution

In [32]:
# Solution

 10. The App Store dataset doesn´t have the ```installs``` information. Then, you will construct the average rating from the ```rating_count_tot``` column per genre. Based on that exercise, give one app profile recommendation for the App Store

In [33]:
# Solution
import pandas as pd

#Primero debemos identificar el nombre de la columna de los "géneros" de las aplicaciones
apps_gratuitas_Apple_Store.head() #prime_genre


id               track_name  size_bytes currency  price  \
0  284882215                 Facebook   389879808      USD    0.0   
1  389801252                Instagram   113954816      USD    0.0   
2  529479190           Clash of Clans   116476928      USD    0.0   
3  420009108               Temple Run    65921024      USD    0.0   
4  284035177  Pandora - Music & Radio   130242560      USD    0.0   

   rating_count_tot  rating_count_ver  user_rating  user_rating_ver      ver  \
0           2974676               212          3.5              3.5     95.0   
1           2161558              1289          4.5              4.0    10.23   
2           2130805               579          4.5              4.5  9.24.12   
3           1724546              3842          4.5              4.0    1.6.2   
4           1126879              3594          4.0              4.5    8.4.1   

  cont_rating        prime_genre  sup_devices.num  ipadSc_urls.num  lang.num  \
0          4+  Social Networking               37                1        29   
1         12+      Photo & Video               37                0        29   
2          9+              Games               38                5        18   
3          9+              Games               40                5         1   
4         12+              Music               37                4         1   

   vpp_lic  En_ingles  
0        1       True  
1        1       True  
2        1       True  
3        1       True  
4        1       True

In [34]:
# Ahora, continuamos con las columnas relevantes se llaman 'rating_count_tot' y 'genre'
promedios_por_genero = apps_gratuitas_Apple_Store.groupby('prime_genre')['rating_count_tot'].mean() #'promedios_por_genero' es una Serie donde el índice son los géneros y los valores son las calificaciones promedio

# Para hacer una recomendación de perfil de aplicación, eligiremos una aplicación del género con la calificación promedio más alta
genero_recomendado = promedios_por_genero.idxmax()

print(f"Recomendaría desarrollar una aplicación en el género '{genero_recomendado}', ya que tiene la calificación promedio más alta.")

Recomendaría desarrollar una aplicación en el género 'Navigation', ya que tiene la calificación promedio más alta.
